### PR TITLE
Dedicated sentry client

### PR DESCRIPTION
**Pull request by dmitryduev**
Fixes WB-9358

Description
-----------

Problem: if `wandb` is imported after a third-party `sentry` client has been set up, the Sentry's global `hub/client` would get overwritten meaning that we would suck in other people's `sentry` events, i.e.:

```python
import sentry_sdk
sentry_sdk.init(...)
import wandb
# `sentry_sdk.Hub.current.client` now points at our Sentry DSN
```

In this PR:
- Use a dedicated Sentry client to ensure we're not interfering with potential third-party usage of Sentry.
- Fixed a bug in `sentry_set_scope` where we potentially would pollute other people's Sentry events with our context info.

Testing
-------
I wrote two functional tests + tested locally/manually using the same script/scenarios with real Sentry DSN and confirm that it worked as expected. (as a side note: I felt it wasn't worth fully mocking logging third-party sentry events and counting their number, so just tested it manually).

Checklist
-------
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,22 +1,27 @@
...
+    rev: v4.2.0
+    hooks:
+      - id: check-yaml
+      - id: end-of-file-fixer
+      - id: trailing-whitespace
+#  - repo: https://github.com/pycqa/isort
+#    rev: 5.5.2
+#    hooks:
+#      - id: isort
+#        args: ["--profile", "black", "--filter-files"]
```
Comment (dmitryduev) [inline .pre-commit-config.yaml:14]: `isort`'s defaults are in conflict with our flake8's defaults, need to address this elsewhere.

Comment (kptkin) [reply .pre-commit-config.yaml:14]: yeah i guess we can drop it for now... until we organize our lint tools

File: wandb/util.py [hunk 2]
```diff
@@ -99,19 +98,23 @@ def get_platform_name() -> str:
...
+    "https://a2f1d701163c42b097b9588e56b1c37e@o151352.ingest.sentry.io/5288891"
+)
 if error_reporting_enabled():
-    default_dsn = (
-        "https://a2f1d701163c42b097b9588e56b1c37e@o151352.ingest.sentry.io/5288891"
-    )
-    sentry_dsn = os.environ.get(SENTRY_DSN, default_dsn)
-    sentry_sdk.init(
+    sentry_dsn = os.environ.get(SENTRY_DSN, sentry_default_dsn)
+    sentry_client = sentry_sdk.Client(
```
Comment (kptkin) [inline wandb/util.py:109]: should we leave a comment here that we are using "internal" classes from `sentry` in the small chance they are going to change it?

File: pyproject.toml [hunk 3]
```diff
@@ -16,3 +16,6 @@ exclude = '''
 | __pycache__
 | .pyc
 '''
+
+[tool.isort]
+profile = "black"
```
Comment (kptkin) [inline pyproject.toml:21]: i guess you want to remove this one as well?

File: functional_tests/debug/sentry-third-party.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        try:
+            print(
+                f"Raising exception #{i + 1} to be captured by third-party sentry client"
+            )
+            raise ValueError("Catch me if you can!")
+        except ValueError:
+            sentry_sdk.capture_exception()
+
+    num_third_party_sentry_events = len(third_party_sentry_events)
+    run.log({"num_third_party_sentry_events": num_third_party_sentry_events})
```
Comment (kptkin) [inline functional_tests/debug/sentry-third-party.py:41]: haha ... that's clever

File: wandb/util.py [hunk 5]
```diff
@@ -99,19 +98,23 @@ def get_platform_name() -> str:
 
 
 # TODO(sentry): This code needs to be moved, sentry shouldn't be initialized as a
-# side effect of loading a module.
+#  side effect of loading a module.
```
Comment (kptkin) [inline wandb/util.py:101]: in the current design if we want to disable sdk_sentry we still need to provide the env variable before the import correct?

Comment (dmitryduev) [reply wandb/util.py:101]: yup

File: wandb/util.py (hunk 2 above)
Comment (dmitryduev) [reply wandb/util.py:109]: they are not really internal, the are public in that they are not underscored and are listed in `__all__`, so I'd leave it as is.

Comment (dmitryduev) [reply wandb/util.py:109]: it's just that their documentation could be more verbose on that. :)

File: wandb/util.py [hunk 6]
```diff
@@ -153,11 +157,12 @@ def sentry_exc(
...
         if isinstance(exc, str):
-            capture_exception(Exception(exc))
+            sentry_hub.capture_exception(Exception(exc))  # type: ignore
         else:
-            capture_exception(exc)
-        if delay:
-            time.sleep(2)
+            sentry_hub.capture_exception(exc)  # type: ignore
+    if delay:
+        time.sleep(2)
```
Comment (dmitryduev) [inline wandb/util.py:164]: @raubitsj, @kptkin and I had a question about this magic. We understand what/why it does, but not all the implications of doing things this way (see, e.g. wandb/sdk/internal/internal.py:L168)

**Comment by dmitryduev** [comment]
For future reference:

Consider the following example script `junk.py`:

```python
import shutil

import wandb

third_party_sentry_events = [1, 2]
run = wandb.init()

print(len(third_party_sentry_events))

num_third_party_sentry_events = len(third_party_sentry_events)
run.log({"num_third_party_sentry_events": num_third_party_sentry_events})

# Triggers a FileNotFoundError from the internal process
# because the internal process reads/writes to the current run directory.
shutil.rmtree(run.dir)
try:
    run.log({"data": 5})
except FileNotFoundError:
    pass
```

Previously, if run with `WANDB_ERROR_REPORTING=false python junk.py`, `num_third_party_sentry_events` won't make it to the summary, whereas without it, `time.sleep(2)` in `sentry_exc(exc, delay=True)` will get triggered and `num_third_party_sentry_events` will be in the run's summary.

**Review by kptkin — APPROVED** [review]
Love to see it... 🤩

**Pull request merged**